src/nn/visit_tensors.rs

```diff
@@ -0,0 +1,177 @@
+use crate::shapes::Dtype;
+
+use crate::prelude::*;
+use std::{
+    fmt::Debug,
+    string::{String, ToString},
+};
+
+pub struct ModuleGroup<'a, const N: usize, const M: usize, T> {
+    pub refs: [&'a T; N],
+    pub refs_mut: [&'a mut T; M],
+    pub name: Option<String>,
+}
+
+// TODO? : Prevent heap allocation using unsafe or ArrayVec
+fn map_mut<'a, A: 'a + Debug, B: 'a + Debug, const N: usize, F: FnMut(&'a mut A) -> &'a mut B>(
+    arr: &'a mut [A; N],
```

Comment:
If it's possible to make this `[&'a mut A; N]` then you can just call map directly

Reply:
This is necessary because `array.map` consumes the original array. We are able to use `array.map` for immutable references because they implement Copy, but mutable references don't implement Copy or Clone.